### PR TITLE
Automatically Update OpenAPI Documentation

### PR DESCRIPTION
**Pull request by confused-Techie**
### Requirements 

* Filling out the template is required.
* All new code requires tests to ensure against regressions.
  - However, if your PR contains zero code changes, feel free to select the checkmark below to indicate so.

* [ ] Have you ran tests against this code?
* [X] This PR contains zero code changes.

### Description of the Change

This PR finally takes advantage of all the work done when modularizing the endpoints and query parameters.

Now we have a system of code objects that define how the code works, define how the tests work, and define the documentation that's generated in a complex way.

This script uses all information available to generate updates to the `OpenAPI` yaml document on each commit, so that we can ensure our schema never becomes out of date

---

Superseedes #22

### CODE REVIEW

**Comment by Spiker985** [comment]
Super excited to see this

**Comment by Spiker985** [comment]
One thing that I noticed on the generated YAML, while not a requirement, is that the endpoints appear to not be alphabetized.
If they were 
https://github.com/pulsar-edit/package-backend/blob/231de711db097dddda06d76095ec4976d5072e52/docs/swagger/openapi3_def.yaml#L71
should be where
https://github.com/pulsar-edit/package-backend/blob/231de711db097dddda06d76095ec4976d5072e52/docs/swagger/openapi3_def.yaml#L17
currently is

But also there are quite a few endpoints that appear to not have any "response" information such as [GET /api/login](https://github.com/pulsar-edit/package-backend/blob/231de711db097dddda06d76095ec4976d5072e52/docs/swagger/openapi3_def.yaml#L20), [GET /api/oauth](https://github.com/pulsar-edit/package-backend/blob/231de711db097dddda06d76095ec4976d5072e52/docs/swagger/openapi3_def.yaml#L25), [GET /api/pat](https://github.com/pulsar-edit/package-backend/blob/231de711db097dddda06d76095ec4976d5072e52/docs/swagger/openapi3_def.yaml#L69) [POST /api/packages](https://github.com/pulsar-edit/package-backend/blob/231de711db097dddda06d76095ec4976d5072e52/docs/swagger/openapi3_def.yaml#L30), [GET /api/packages/{packageName}/versions/{versionName}](https://github.com/pulsar-edit/package-backend/blob/231de711db097dddda06d76095ec4976d5072e52/docs/swagger/openapi3_def.yaml#L153)

The Tarball endpoint response is a little screwy too, though I assume that's an indexing problem: https://github.com/pulsar-edit/package-backend/blob/231de711db097dddda06d76095ec4976d5072e52/docs/swagger/openapi3_def.yaml#L229-L231
similar problem here as well
https://github.com/pulsar-edit/package-backend/blob/231de711db097dddda06d76095ec4976d5072e52/docs/swagger/openapi3_def.yaml#L279-L282

**Comment by confused-Techie** [comment]
@Spiker985 Thanks a ton for pointing this out!

Not sure what's gone wrong with the tarball endpoint so I'll take a look. But I think the other issues is due to those endpoints not being fully written out, as in the object lacks details of it's response data. So I'll make sure to add that so we can start testing for it as well.

Otherwise you're right that alphabetizing this would be really cool, I almost wonder, I'm pretty sure the `yaml.dumb()` method supports a custom sort function, maybe adding it there would be enough to do that.

But thanks for reviewing the document, since wanna make sure it's valid and keeps up the standard of what your originally wrote

**Comment by confused-Techie** [comment]
@Spiker985 Would love to hear any feedback, I think I've addressed everything. From here there's just the case of a few missing responses and parameters

File: docs/swagger/openapi3_def.yaml [hunk 1]
```diff
@@ -1,621 +1,732 @@
...
+        description: Owner of packages to retreive.
+        example: pulsar-edit
+        in: path
+        name: ownerName
+        required: true
+        schema:
+          type: string
+    packageName:
+      allowEmptyValue: false
+      description: The name of the package to reutrn details for. Must be URL escaped.
```
Comment (Daeraxa) [inline docs/swagger/openapi3_def.yaml:75]:
```suggestion
      description: The name of the package to return details for. Must be URL escaped.
```

File: docs/swagger/openapi3_def.yaml [hunk 2]
```diff
@@ -1,621 +1,732 @@
...
+      description: The service of which to filter packages by
+      example: autocomplete.watchEditor
+      in: query
+      name: service
+      schema:
+        type: string
+    serviceType:
+      allowEmptyValue: true
+      deescription: >-
+        Choos whether to display 'consumer' or 'providers' of the specified
```
Comment (Daeraxa) [inline docs/swagger/openapi3_def.yaml:130]:
```suggestion
        Choose whether to display 'consumer' or 'providers' of the specified
```

File: docs/swagger/openapi3_def.yaml [hunk 3]
```diff
@@ -1,621 +1,732 @@
-openapi: 3.0.1
+components:
+  parameters:
+    auth:
+      allowEmptyValue: false
+      description: Authorization Headers
```
Comment (Daeraxa) [inline docs/swagger/openapi3_def.yaml:5]:
```suggestion
      description: Authorization Headers.
```

File: docs/swagger/openapi3_def.yaml [hunk 4]
```diff
@@ -1,621 +1,732 @@
...
+        Only show packages compatible with this Pulsar version. Must be a valid
+        Semver.
+      example: 1.0.0
+      in: query
+      name: engine
+      schema:
+        type: string
+    fileExtension:
+      allowEmptyValue: true
+      description: File extension of which to only show compatible grammar package's of.
```
Comment (Daeraxa) [inline docs/swagger/openapi3_def.yaml:23]:
```suggestion
      description: File extension of which to only show compatible grammar packages of.
```

File: docs/swagger/openapi3_def.yaml [hunk 5]
```diff
@@ -1,621 +1,732 @@
...
+      required: false
+      schema:
+        default: package
+        enum:
+          - package
+          - theme
+        type: string
+    login:
+      allowEmptyValue: false
+      description: The User from the URL Path
```
Comment (Daeraxa) [inline docs/swagger/openapi3_def.yaml:46]:
```suggestion
      description: The User from the URL Path.
```

File: docs/swagger/openapi3_def.yaml [hunk 6]
```diff
@@ -1,621 +1,732 @@
...
+        description: Owner to filter results by.
+        example: pulsar-edit
+        in: query
+        name: owner
+        required: false
+        schema:
+          type: string
+      ownerName:
+        allowEmptyValue: false
+        description: Owner of packages to retreive.
```
Comment (Daeraxa) [inline docs/swagger/openapi3_def.yaml:66]:
```suggestion
        description: Owner of packages to retrieve.
```

File: docs/swagger/openapi3_def.yaml [hunk 7]
```diff
@@ -1,621 +1,732 @@
...
+      description: Repository to publish.
+      example: pulsar-edit/pulsar
+      in: query
+      name: repository
+      required: true
+      schema:
+        type: string
+    service:
+      allowEmptyValue: true
+      description: The service of which to filter packages by
```
Comment (Daeraxa) [inline docs/swagger/openapi3_def.yaml:121]:
```suggestion
      description: The service of which to filter packages by.
```

File: docs/swagger/openapi3_def.yaml [hunk 8]
```diff
@@ -1,621 +1,732 @@
...
+      in: query
+      name: serviceType
+      schema:
+        enum:
+          - consumed
+          - provided
+        type: string
+    serviceVersion:
+      allowEmptyValue: true
+      description: Filter by a specific version of the 'service'
```
Comment (Daeraxa) [inline docs/swagger/openapi3_def.yaml:142]:
```suggestion
      description: Filter by a specific version of the 'service'.
```

File: docs/swagger/openapi3_def.yaml [hunk 9]
```diff
@@ -1,621 +1,732 @@
...
+      description: TODO
+      example: TODO
+      in: query
+      name: tag
+      required: false
+      schema:
+        type: string
+    versionName:
+      allowEmptyValue: false
+      description: The version of the package to access
```
Comment (Daeraxa) [inline docs/swagger/openapi3_def.yaml:175]:
```suggestion
      description: The version of the package to access.
```

File: docs/swagger/openapi3_def.yaml [hunk 10]
```diff
@@ -1,621 +1,732 @@
...
+      example: 1.0.0
+      in: path
+      name: versionName
+      required: true
+      schema:
+        type: string"
+  schemas:
+    message:
+      description: >-
+        A generic object that could contain status information or error
```
Comment (Daeraxa) [inline docs/swagger/openapi3_def.yaml:185]:
```suggestion
        A generic object that could contain status information or error.
```

File: docs/swagger/openapi3_def.yaml [hunk 11]
```diff
@@ -1,621 +1,732 @@
...
-          $ref: "#/components/responses/404"
-        500:
-          $ref: "#/components/responses/500"
-  /api/packages/{packageName}/star:
+        '200':
+          content:
+            application/json:
+              schema:
+                $ref: '#/components/schemas/packageObjectFull'
+          description: A 'Package Object Full' of the modified package
```
Comment (Daeraxa) [inline docs/swagger/openapi3_def.yaml:464]:
```suggestion
          description: A 'Package Object Full' of the modified package.
```

File: docs/swagger/openapi3_def.yaml [hunk 12]
```diff
@@ -1,621 +1,732 @@
...
+      responses:
+        '200':
+          content: {}
+          description: Atom update feed, following the format expected by Squirrel.
+      summary: List Pulsar Updates
+  /api/users:
+    get:
+      description: >-
+        This endpoint only exists on the web version of the upstream API. Having
+        no backend equivolent.
```
Comment (Daeraxa) [inline docs/swagger/openapi3_def.yaml:683]:
```suggestion
        no backend equivalent.
```

File: scripts/openapi/main.js [hunk 13]
```diff
@@ -0,0 +1,188 @@
...
+      description: "Production Server"
+    },
+    {
+      url: "http://localhost:8080",
+      description: "Locally hosted development server"
+    }
+  ],
+};
+
+// With our initial setup done, lets add our paths onto the schema
```
Comment (Daeraxa) [inline scripts/openapi/main.js:50]:
```suggestion
// With our initial setup done, let's add our paths onto the schema
```

**Comment by Daeraxa** [comment]
Sorry, should have done it as a single review, thought it was just one or two then I went a bit mad... Some are typos, many are just adding a full stop to the end of the sentence to match the style of the others.

**Comment by confused-Techie** [comment]
@Daeraxa You're feedback has been addressed, thanks a ton for the review and taking a look at this

**Comment by confused-Techie** [comment]
@Spiker985 @Daeraxa Alright, from here I've addressed the issues that had still existed (as well as discovering another) and I think we are 100% ready to merge if anyone wants to take a last look.

As for what I've addressed in the last couple PRs:
* This PR would clobber entries with the same path, such as `GET /api/packages` and `POST /api/packages` so I've resolved that
* Got rid of `multiSchema` idea, and instead have a `shared` dir for parameters that have the same handling but need unique schemas
* And resolved the issues around a module having a different `components` entry because of a mismatch between the name and module name

**Comment by Spiker985** [comment]
Nitpicky, but I've never seen any OpenAPI schemas which list the components first

Does this have any effect on how the swagger page displays them?

**Comment by confused-Techie** [comment]
> Nitpicky, but I've never seen any OpenAPI schemas which list the components first
> 
> Does this have any effect on how the swagger page displays them?

Valid question, threw me off at first too.

But doesn't seem to have any negative effect: 

image

But this likely happens because we now sort the results to list the endpoints in a more satisfying way. If we really wanted to I could try my hand at a custom sorter, but as it doesn't seem to affect anything I don't think we have anything to worry about.

**Comment by Spiker985** [comment]
I'll likely reread this PR later today just to double check and make sure I haven't missed anything with the other changes

But if it gets a green light before then, feel free to merge

Arguably, the "current" version is no more accurate than anything that would be missed in this PR anyway

**Comment by Daeraxa** [comment]
> Nitpicky, but I've never seen any OpenAPI schemas which list the components first

Do you have an example of what you normally expect? It looks the same as the existing UI from what I can see or am I missing the issue here?

**Comment by Spiker985** [comment]
> > Nitpicky, but I've never seen any OpenAPI schemas which list the components first
> 
> @Spiker985 Do you have an example of what you normally expect? It looks the same as the existing UI from what I can see or am I missing the issue here?

Nope, that was exactly what I wanted to be sure of.

I was unsure of how the OpenAPI UI implementation was affected by the schema

From confused's response, it appears that the schema ordering is irrelevant and the UI will always display in a "static" order of API information -> endpoints -> schema

Basically, as long as the data exists in your yaml file, regardless of order, it'll always be implemented the same in the UI

That was all I wanted to validate with that question

(This doesn't count as reading through the updates, because I haven't done that yet)

**Comment by confused-Techie** [comment]
Alright this PR now has `OPTIONS` methods added, as well as everything else should be fully fleshed out

**Comment by confused-Techie** [comment]
Alright, since it's been a week since the last comment here, and this PR touches far too much of the codebase to avoid future conflicts, I'm going to go ahead and merge.

Thanks a ton to those that helped get this one along!

**Pull request merged**